File: pdl/infra/kotlin/src/main/kotlin/io/pdl/infrastructure/data/StoreRegistry.kt
package io.pdl.infrastructure.data

object StoreRegistry {
    private var defaultStore: DBStore? = null

    fun setDefaultStore(store: DBStore) {
        defaultStore = store
    }

    fun defaultStore(): DBStore = defaultStore
        ?: error("io.pdl.infrastructure.data: default store not configured")

    fun resolve(override: DBStore?): DBStore = override ?: defaultStore()
}

File: pdl/infra/php/src/Io/Pdl/Infrastructure/Data/StoreRegistry.php
<?php

namespace Io\Pdl\Infrastructure\Data;

use RuntimeException;

final class StoreRegistry
{
    private static ?DBStore $defaultStore = null;

    public static function setDefaultStore(DBStore $store): void
    {
        self::$defaultStore = $store;
    }

    public static function defaultStore(): DBStore
    {
        if (self::$defaultStore === null) {
            throw new RuntimeException('Io\Pdl\Infrastructure\Data: default store is not configured');
        }

        return self::$defaultStore;
    }

    public static function resolve(?DBStore $store): DBStore
    {
        return $store ?? self::defaultStore();
    }
}
